### PR TITLE
SCMOD-4887: Record poison messages

### PR DESCRIPTION
**Pull request by kusumaghoshdastidar**
Jira: https://portal.digitalsafe.net/browse/SCMOD-4887

### CODE REVIEW

**Review by rorytorneymf — APPROVED** [review]

**Review by Anthony-McGreevy — APPROVED** [review]

File: worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java [hunk 1]
```diff
@@ -233,4 +234,34 @@ public void setPriority(Integer priority)
...
+        builder.append("TaskMessage [version=");
+        builder.append(version);
+        builder.append(", taskId=");
+        builder.append(taskId);
+        builder.append(", taskClassifier=");
+        builder.append(taskClassifier);
+        builder.append(", taskApiVersion=");
+        builder.append(taskApiVersion);
+        builder.append(", taskData=");
+        builder.append(Arrays.toString(taskData));
```
Comment (andyreidz) [inline worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java]: taskData is a byte array, I think Arrays.toString will corrupt the taskData.

File: worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java [hunk 2]
```diff
@@ -48,8 +48,11 @@ public void run()
     {
         try {
             if (workerTask.isPoison()) {
-                LOG.warn("Worker [" + worker.getWorkerIdentifier() + "] did not handle poisoned message, when it was passed for processing.");
-                throw new RuntimeException("Worker [" + worker.getWorkerIdentifier() + "] did not handle poisoned message, when it was passed for processing.");
+                final String message = "Worker [" + worker.getWorkerIdentifier()
+                + "] did not handle poisoned message, when it was passed for processing. "
+                + workerTask.getTaskMessage();
+                LOG.warn(message);
```
Comment (andyreidz) [inline worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java]: The message contains the taskData which could contain sensitive information from the user.
The taskData could also be quite big.

**Review by andyreidz — CHANGES_REQUESTED** [review]

File: worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java (hunk 2 above)
Comment (kusumaghoshdastidar) [reply worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java]: I removed taskData from the log but included it in the exception message. Without the taskData not sure if the message would have sufficient details to understand what caused the failure.

Comment (andyreidz) [reply worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java]: Great, it was the logging line above the comment was related to.
image

File: worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java (hunk 1 above)
Comment (andyreidz) [reply worker-api/src/main/java/com/hpe/caf/api/worker/TaskMessage.java]: I saw you updated the test to get the bytes for a string, was that trying to show there was no problem?

File: worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java [hunk 3]
```diff
@@ -48,8 +48,11 @@ public void run()
...
+                throw new RuntimeException(message + " TaskData: " + workerTask.getData());
```
Comment (andyreidz) [inline worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java]: How will this behave with binary data?

File: worker-core/src/test/java/com/hpe/caf/worker/core/StreamingWorkerWrapperTest.java [hunk 4]
```diff
@@ -212,6 +212,40 @@ public void testAbandon()
...
+        String queueMsgId = "exception";
+        CountDownLatch latch = new CountDownLatch(1);
+        TestCallback callback = new TestCallback(latch);
+        TaskMessage m = new TaskMessage();
+        ServicePath path = new ServicePath(SERVICE_NAME);
+        Map<String, byte[]> contextMap = new HashMap<>();
+        contextMap.put(path.toString(), SUCCESS_BYTES);
+        m.setTaskId(TASK_ID);
+        m.setContext(contextMap);
+        m.setTaskData("Test data".getBytes(StandardCharsets.UTF_8));
```
Comment (andyreidz) [inline worker-core/src/test/java/com/hpe/caf/worker/core/StreamingWorkerWrapperTest.java:241]: The taskdata might not be a byte representation of a string (which would always contain printable characters)
Test this with more challenging binary data (and let me know if there is a problem or not)

File: worker-core/src/test/java/com/hpe/caf/worker/core/StreamingWorkerWrapperTest.java [hunk 5]
```diff
@@ -212,6 +212,40 @@ public void testAbandon()
...
+        Map<String, Object> headers = new HashMap<>();
+        WorkerTaskImpl workerTask = new WorkerTaskImpl(path, callback, happyWorkerFactory, getMockTaskInformation(queueMsgId), m, true,
+                headers, codec, priorityManager);
+        StreamingWorkerWrapper wrapper = new StreamingWorkerWrapper(workerTask);
+        Thread t = new Thread(wrapper);
+        t.start();
+        latch.await(5, TimeUnit.SECONDS);
+        String s = codec.deserialise(callback.getResultData(), String.class);
```
Comment (andyreidz) [inline worker-core/src/test/java/com/hpe/caf/worker/core/StreamingWorkerWrapperTest.java]: Hint to my point on converting byte array to string.

**Review by andyreidz — CHANGES_REQUESTED** [review]

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
As discussed, wrap the poison message and output it to worker's error queue using the `workerTask.sendMessage()` method.

File: worker-api/src/main/java/com/hpe/caf/api/worker/WorkerFactory.java [hunk 6]
```diff
@@ -62,6 +62,11 @@ default WorkerConfiguration getWorkerConfiguration()
      */
     String getInvalidTaskQueue();
 
+    /**
+     * @return the queue publish poison messages
+     */
+    String getRejectedTaskQueue();
```
Comment (dermot-hardy) [inline worker-api/src/main/java/com/hpe/caf/api/worker/WorkerFactory.java]: This is a breaking change that would require a series of workers to be updated.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java [hunk 7]
```diff
@@ -765,4 +765,21 @@ private static int getBufferLimit()
         return bufferLimit;
     }
 
+    void sendPoisonMessage(final String workerIdentifier)
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java]: - Why is this package-private?
- Why is it here?  Can't it just be a private utility function in `StreamingWorkerWrapper`?

File: worker-archetype/src/main/resources/archetype-resources/__rootArtifactId__-container/src/main/config/cfg~caf~worker~__workerName__Configuration.js [hunk 8]
```diff
@@ -18,6 +18,8 @@
     workerVersion: "${project.version}",
     outputQueue: getenv("CAF_WORKER_OUTPUT_QUEUE")
             || (getenv("CAF_WORKER_BASE_QUEUE_NAME") || getenv("CAF_WORKER_NAME") || "worker") + "-out",
+    rejectQueue: getenv("CAF_WORKER_REJECT_QUEUE")
+            || (getenv("CAF_WORKER_BASE_QUEUE_NAME") || getenv("CAF_WORKER_NAME") || "worker") + "-reject",
```
Comment (dermot-hardy) [inline worker-archetype/src/main/resources/archetype-resources/__rootArtifactId__-container/src/main/config/cfg~caf~worker~__workerName__Configuration.js:22]: It's fine to change the archetype but don't forget that it only effects future workers.  For existing workers which haven't had this line added `rejectQueue` will be `null`.  The code will have to deal with this.

File: worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java [hunk 9]
```diff
@@ -76,4 +82,20 @@ public static Timer getTimer()
...
+    {
+        // Publish poison message to "reject" queue
+        final TaskMessage poisonMessage = new TaskMessage(
+                UUID.randomUUID().toString(),
+                MoreObjects.firstNonNull(workerTask.getClassifier(), ""),
+                workerTask.getVersion(),
+                workerTask.getData(),
+                TaskStatus.RESULT_EXCEPTION,
+                Collections.<String, byte[]>emptyMap(),
+                workerTask.getRejectQueue(),
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/StreamingWorkerWrapper.java:96]: Per my previous comment, what if this is `null`?

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**